dill/_dill.py

```diff
@@ -509,20 +559,42 @@ def dump_session(filename='/tmp/session.pkl', main=None, byref=False, **kwds):
             file.close()
     return
 
-def _open_peekable(filename):
-    if hasattr(filename, 'read'):
-        file = filename
-    else:
-        file = open(filename, 'rb')
-    if not hasattr(file, 'peek'):
+class _PeekableReader:
+    """lightweight stream wrapper that implements peek()"""
+    def __init__(self, stream):
+        self.stream = stream
+    def read(self, n):
+        return self.stream.read(n)
+    def tell(self):
+        return self.stream.tell()
+    def close(self):
+        return self.stream.close()
+    def peek(self, n):
+        stream = self.stream
         try:
-            import io
-            file = io.BufferedReader(file)
+            if hasttr(stream, 'flush'): stream.flush()
+            position = stream.tell()
+            stream.seek(position)  # assert seek() works before reading
+            chunk = stream.read(n)
+            stream.seek(position)
+            return chunk
+        except (AttributeError, OSError):
+            raise NotImplementedError("stream is not peekable: %r", stream) from None
+
+def _make_peekable(stream):
+    """return stream as an object with a peek() method"""
+    import io
+    if hasattr(stream, 'peek'):
+        return stream
+    if not (hasattr(stream, 'tell') and hasattr(stream, 'seek')):
+        try:
+            return io.BufferedReader(stream)
         except Exception:
-            pass  # ...and hope for the best
-    return file
+            pass
+    return _PeekableReader(stream)
 
-def _inspect_pickle(file, main_is_none):
+def _identify_session_module(file, main_is_none):
```

Comment:
Ok, this looks good. I'm going to ask you to think about naming again. How about `_identify_main` or `_identify_main_module` or `_identify_module`?  Also, the arg name `main_is_none` is not good... what is the intent of this arg?

Reply:
Rewriting comment: I prefer `_identify_module()`. And `main_is_none` was a poor choice. Changed it.